x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java

```diff
@@ -81,12 +83,62 @@ public Object extract(SearchHit hit) {
         Object value = null;
         DocumentField field = null;
         if (hitName != null) {
-            // a nested field value is grouped under the nested parent name (ie dep.dep_name lives under "dep":[{dep_name:value}])
-            field = hit.field(hitName);
+            value = extractNestedField(hit);
         } else {
             field = hit.field(fieldName);
+            if (field != null) {
+                value = unwrapFieldsMultiValue(field.getValues());
+            }
+        }
+        return value;
+    }
+
+    /*
+     * For a tree of fields like root.nested1.nested2.leaf where nested1 and nested2 are nested field types,
+     * fieldName is root.nested1.nested2.leaf, while hitName is root.nested1.nested2
+     * We first look for root.nested1.nested2 or root.nested1 or root in the SearchHit until we find something.
+     * If the DocumentField lives under "root.nested1" the remaining path to search for (in the DocumentField itself) is nested2.
+     * After this step is done, what remains to be done is just getting the leaf values.
+     */
+    @SuppressWarnings("unchecked")
+    private Object extractNestedField(SearchHit hit) {
+        Object value;
+        DocumentField field;
+        String tempHitname = hitName;
+        List<String> remainingPath = new ArrayList<>();
+        // first, search for the "root" DocumentField under which the remaining path of nested document values is
+        while((field = hit.field(tempHitname)) == null) {
+            int indexOfDot = tempHitname.lastIndexOf(".");
```

Comment:
Is this guaranteed to return a non-negative?
Actually, with a mapping similar to the  exemplified one in the test's comment (but only using the first lowered letter of the randomly generated nodes), indexing just `{"h": [{"f":{"b": {"a": "true"}}}]}` and querying with `SELECT  h.f.o FROM test` seems to trigger:
```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 1
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3734)
	at java.base/java.lang.String.substring(String.java:1903)
	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extractNestedField(AbstractFieldHitExtractor.java:113)
	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extract(AbstractFieldHitExtractor.java:86)
	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:129)
	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:32)
	at org.elasticsearch.xpack.sql.execution.search.ResultRowSet.getColumn(ResultRowSet.java:38)
	at org.elasticsearch.xpack.sql.session.AbstractRowSet.column(AbstractRowSet.java:19)
	at org.elasticsearch.xpack.sql.session.RowView.forEachColumn(RowView.java:39)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$createResponse$5(TransportSqlQueryAction.java:132)
	at org.elasticsearch.xpack.sql.session.RowSet.forEachRow(RowSet.java:28)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponse(TransportSqlQueryAction.java:130)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponseWithSchema(TransportSqlQueryAction.java:125)
	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$operation$0(TransportSqlQueryAction.java:97)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:139)

```

Reply:
Mentioning it here, but happy to open an issue if better (or drop it, if wrong): selecting `h.f.b.a` will still return nothing.
Reason being that the `inner_hits` response from ES comes with with the entries listed as `h.f_12` (for example), while we're looking for the inner hit `h.f.b` in SearchHitRowSet. Seems like this should be doable to work, but didn't look deeper.
This is not new, this didn't work before either (although no error message is returned).

Reply:
Nice catch! That's an older problem, for sure. It's the complexity of the `nested` documents definition possibilities that shows that our algorithm isn't covering all of them. I'd be inclined in dealing with this bug in a subsequent PR as it doesn't look straightforward and it may need a bit of rethinking the way we deal with `nested` fields hierarchies.

Reply:
> Is this guaranteed to return a non-negative?
> Actually, with a mapping similar to the exemplified one in the test's comment (but only using the first lowered letter of the randomly generated nodes), indexing just `{"h": [{"f":{"b": {"a": "true"}}}]}` and querying with `SELECT h.f.o FROM test` seems to trigger:
> 
> ```
> Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 1
> 	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3734)
> 	at java.base/java.lang.String.substring(String.java:1903)
> 	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extractNestedField(AbstractFieldHitExtractor.java:113)
> 	at org.elasticsearch.xpack.ql.execution.search.extractor.AbstractFieldHitExtractor.extract(AbstractFieldHitExtractor.java:86)
> 	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:129)
> 	at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:32)
> 	at org.elasticsearch.xpack.sql.execution.search.ResultRowSet.getColumn(ResultRowSet.java:38)
> 	at org.elasticsearch.xpack.sql.session.AbstractRowSet.column(AbstractRowSet.java:19)
> 	at org.elasticsearch.xpack.sql.session.RowView.forEachColumn(RowView.java:39)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$createResponse$5(TransportSqlQueryAction.java:132)
> 	at org.elasticsearch.xpack.sql.session.RowSet.forEachRow(RowSet.java:28)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponse(TransportSqlQueryAction.java:130)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.createResponseWithSchema(TransportSqlQueryAction.java:125)
> 	at org.elasticsearch.xpack.sql.plugin.TransportSqlQueryAction.lambda$operation$0(TransportSqlQueryAction.java:97)
> 	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:139)
> ```

We should prevent the error from occurring - looking at the stacktrace this seems to have been addressed yet I'm mentioning to double check this is indeed the case.

Reply:
@bpintea [this change](https://github.com/elastic/elasticsearch/pull/69368/files#diff-85f0eb71298942313786f64c1d08628baec44a09fcd1c70fa1ee46a4dfad335aR341) should fix the `h.f.b.a` returning nothing example you mentioned. Test for this exact case [here](https://github.com/elastic/elasticsearch/pull/69368/files#diff-990981c3256841bd9e0cc66efff91f49397139b1ae9cf548a8a6cae8c71fe5fdR1139).